question21/main.js

```diff
@@ -0,0 +1,159 @@
+  userId: "ID",
+  name: "名前",
+  gender: "性別",
+  age: "年齢"
+}
+
+const renderTabaleElement = (userData)  => {
+  const table = createElementWithClassName("table", "table");
+  const tbody = createElementWithClassName("tbody", "tbody");
+  table.id = "js-table";
+  tbody.id = "js-tbody";
+  const tableHeadValue = Object.values(userDataColumn);
+  wrapper.appendChild(table).appendChild(getTableHeadElement(tableHeadValue));
+  wrapper.appendChild(table).appendChild(tbody).appendChild(getTableRowFragment(userData));
+  sortButtonClickEvent(userData);
+}
+
+const getTableHeadElement = (tableHeadValue) => {
+  const thead = createElementWithClassName("thead", "thead");
+  const tr = createElementWithClassName("tr", "tr");
+  const fragmentThElement = document.createDocumentFragment();
+  for (const value of tableHeadValue) {
+    const th = createElementWithClassName("th", "th");
+    th.textContent = value;
+    value === 'ID' && th.appendChild(createSortButton());
+    fragmentThElement.appendChild(tr).appendChild(th);
+  }
+  thead.appendChild(fragmentThElement);
+  return thead;
+}
+
+const getTableRowFragment = (userData) => {
+  const tableRowfragment = document.createDocumentFragment();
+  for (let i = 0; i < userData.length; i++) {
+    const tr = createElementWithClassName("tr", "tr");
+    tableRowfragment.appendChild(tr).appendChild(getTableDataFragment(userData[i]));
+  }
+  return tableRowfragment;
+}
+
+const getTableDataFragment = (userData) => {
+  const tableDatafragment = document.createDocumentFragment();
+  Object.keys(userDataColumn).forEach((key) => {
+    const td = createElementWithClassName("td", "td");
+    td.textContent = userData[key];
+    tableDatafragment.appendChild(td);
+  });
+  return tableDatafragment;
+};
+
+const createSortButton = () => {
+  const sortButton = createElementWithClassName("button", "sortButton");
+  sortButton.dataset.status = "default";
+  sortButton.id = "js-sortbtn";
+  return sortButton;
+}
+
+const sortButtonClickEvent = (userData) => {
+  const tbody = document.querySelector('tbody');
+  const sortButton = document.getElementById('js-sortbtn');
+  sortButton.addEventListener('click', (e) => {
+    const nextStatus = getNextStatus(e.target);
+    sortButton.dataset.status = nextStatus;
+    const sortResultRows = sortFunc(e.target,userData);
+    while (tbody.firstChild) {
+      tbody.removeChild(tbody.firstChild);
+  }
+    sortResultRows.forEach((row) => {
+      tbody.appendChild(row);
+    });
+  });
+}
+
+const getNextStatus = (target) => {
+  switch (target.dataset.status) {
+    case 'default':
+      return 'asc';
+    case 'asc':
+      return 'desc';
+    case 'desc':
+      return 'default'; 
+    default:
+      return 'default';
+  }
+}
+
+const sortFunc = (target,userData) => {
+  const trElement = [...document.querySelectorAll('tbody > tr')];
+  const currentStatus = target.dataset.status;
+  const initialTrArray = [];
+  initialTrArray.push(getTableRowFragment(userData)); 
```

Comment:
IMO
I feel that it would be better to receive it as an argument than to create the RowFragment again here, what do you think?


ex)

```js
const sortFunc = (target, defaultRows) => {
  const trElement = [...defaultRows];
  const currentStatus = target.dataset.status;
...
```

```js
const sortButtonClickEvent = () => {
  const tbody = document.querySelector("tbody");
  const sortButton = document.getElementById("js-sortbtn");
  
  // here
  const trElement = [...document.querySelectorAll("tbody > tr")];

  sortButton.addEventListener("click", (e) => {
    const nextStatus = getNextStatus(e.target);
    sortButton.dataset.status = nextStatus;
    const sortResultRows = sortFunc(e.target, trElement);
...
```

Reply:
Thank you for your review!
I was struggling with how to implement this point, but when you mentioned it, I realized it would be cleaner to write that way. I also became aware of points that I did not understand. fixed so please confirm.
fixed commit : https://github.com/csakyo/homework/pull/36/commits/75603e7f11476f1dfb40a7e40ea0024a141c1ee8